### PR TITLE
enosys: add support for ioctl blocking

### PR DESCRIPTION
**Pull request by t-8ch**
Ioctls are blocked with ENOTTY.

This is still missing a mechanism to discover ioctls at build time.
At the moment only `FIOCLEX` is supported.

### CODE REVIEW

**Comment by masatake** [comment]
I guess you can get many hints in the source code of strace; it knows ioctl very well!

I have never tried but I have seen "--inject" option on the man page of strace.

**Comment by t-8ch** [comment]
It seems `strace --inject` and `strace --fault` can actually replace the core functionality of `enosys`.
I'll have to look into this some more, thanks for the pointer!

**Comment by karelzak** [comment]
Ah, it seems enosys is not such an original idea ;-(((

**Comment by t-8ch** [comment]
> Ready? ;-)

Nope :-)

Two problems:

* Finding a list of ioctls from the header files
* When generating code for all ioctls as it's currently done for syscalls (and in this PR) would exceed the maximum size of a BPF program. Will need to make the logic more dynamic.

**Comment by karelzak** [comment]
Maybe we don't have to support all ioctls by name, but we can support a subset of ioctls by name and the rest by number.

   enosys --ioctl <name|hex, ...>

and do not support a program that blocks all ioctls. IMHO, supporting ioctl numbers would be a nice fallback solution for 3rd party stuff (or kernel developers who work on new stuff, etc).

**Comment by t-8ch** [comment]
> we can support a subset of ioctls by name and the rest by number

Sounds good. But that also needs changes to the BPF generator to support dynamic checks.
And then we can also block syscalls by number.

So the development process would look like this:
1) dynamic BPF generation
1) syscalls by number
1) ioctls

**Comment by t-8ch** [comment]
This is ready for review. All planned steps are implemented.
The uninitialized local variables warnings found by CI seem to be false positives.

**Comment by karelzak** [comment]
It looks good to me.

File: misc-utils/enosys.c [hunk 1]
```diff
@@ -55,12 +58,22 @@ struct syscall {
...
 static const struct syscall syscalls[] = {
 #define UL_SYSCALL(name, nr) { name, nr },
 #include "syscalls.h"
 #undef UL_SYSCALL
 };
-static_assert(sizeof(syscalls) > 0, "no syscalls found");
+
+static const struct syscall ioctls[] = {
+	{ "FIOCLEX", FIOCLEX },
+};
```
[the remark this answers is not in the document]
Comment (t-8ch) [reply misc-utils/enosys.c:76]: Surely :-)

I have no idea on how to decide which ones to add though. Any suggestions?

**Pull request merged**